### ISSUE TITLE
Generalize `Ptr` invariant relaxation

### ISSUE THREAD

**Issue by jswrenn** [google/zerocopy#1167]
*WARNING: This might be unsound per #1866.*

*This builds on machinery introduced in #1166.*

It would be nice to introduce a general invariant relaxation method in the form:

```rust
fn relax_invariants<J>(self) -> Ptr<'a, T, J>
where
    I: invariants::AtLeast<J>,
{
    // SAFETY: `I` is at least as restrictive as `J`.
    unsafe { Ptr::from_ptr(self) }
}
```

...and call it in our specific `.forget_*` methods.

We think this might be blocked on stabilizing [`feature(marker_trait_attr)`](https://doc.rust-lang.org/beta/unstable-book/language-features/marker-trait-attr.html) (https://github.com/rust-lang/rust/issues/29864).

**Comment by joshlf** [comment]
Closing in favor of https://github.com/google/zerocopy/issues/1866, which goes in a different direction.
